core-contracts/contracts/bns.clar

```diff
@@ -0,0 +1,958 @@
+        (>= block-height (+ namespace-preorder-claimability-ttl
+                            (expects! (get created-at former-preorder) (err err-panic)))))
+      (err err-namespace-preorder-already-exists))
+          (asserts! (> stx-to-burn u0) (err err-namespace-stx-burnt-insufficient))
+    ;; Ensure that the hashed namespace is 20 bytes long
+    (asserts! (eq? (len hashed-salted-namespace) u20) (err err-namespace-hash-malformed))
+    ;; Ensure that user will be burning a positive amount of tokens
+    (asserts! (> stx-to-burn u0) (err err-namespace-stx-burnt-insufficient))
+    ;; Burn the tokens - todo(ludo): switch to native STX once available
+    (expects! (ft-transfer! stx stx-to-burn contract-caller burn-address) (err err-insufficient-funds))
+    ;; Register the preorder
+    (map-set! namespace-preorders
+      ((hashed-salted-namespace hashed-salted-namespace) (buyer contract-caller))
+      ((created-at block-height) (claimed 'false) (stx-burned stx-to-burn)))
+    (ok (+ block-height namespace-preorder-claimability-ttl))))
+
+;; NAMESPACE_REVEAL
+;; This second step reveals the salt and the namespace ID (pairing it with its NAMESPACE_PREORDER). It reveals how long
+;; names last in this namespace before they expire or must be renewed, and it sets a price function for the namespace
+;; that determines how cheap or expensive names its will be.
+(define-public (namespace-reveal (namespace (buff 19))
+                                 (namespace-version uint)
+                                 (namespace-salt (buff 20))
+                                 (p-func-base uint)
+                                 (p-func-coeff uint)
+                                 (p-func-b1 uint)
+                                 (p-func-b2 uint)
+                                 (p-func-b3 uint)
+                                 (p-func-b4 uint)
+                                 (p-func-b5 uint)
+                                 (p-func-b6 uint)
+                                 (p-func-b7 uint)
+                                 (p-func-b8 uint)
+                                 (p-func-b9 uint)
+                                 (p-func-b10 uint)
+                                 (p-func-b11 uint)
+                                 (p-func-b12 uint)
+                                 (p-func-b13 uint)
+                                 (p-func-b14 uint)
+                                 (p-func-b15 uint)
+                                 (p-func-b16 uint)
+                                 (p-func-non-alpha-discount uint)
+                                 (p-func-no-vowel-discount uint)
+                                 (renewal-rule uint)
+                                 (namespace-import principal))
+  ;; The salt and namespace must hash to a preorder entry in the `namespace_preorders` table.
+  ;; The sender must match the principal in the preorder entry (implied)
+  (let (
+    (hashed-salted-namespace (hash160 (concat namespace namespace-salt)))
```

Comment:
I think we need something like a separator between the namespace and the salt, e.g.,

these two (namespace, salts) would have the same salted-hash:

```
1: 'abc', 'def'
2: 'ab', 'cdef'
```